File: pkg/util/bayes/negative_binomial.R
# Notes:
# Rscript <filename> will run this.  It only uses the standard library.

# To detect problems faster can:
# * We can use smaller priors so new data has a stronger influence.
# * Calculate posterior predictive distributions to detect deviations from what we expect.

# This is where we mix in the historical and recent data.
posterior_params <- function(alpha_prior, beta_prior, k_recent, n_recent) {
  alpha_posterior <- alpha_prior + k_recent
  beta_posterior <- beta_prior + n_recent
  return(list(alpha = alpha_posterior, beta = beta_posterior))
}

posterior_predictive <- function(alpha_posterior, beta_posterior, k, n) {
  prob <- dnbinom(x = k, size = alpha_posterior, prob = beta_posterior / (beta_posterior + n))
  return(prob)
}

analyze_scenario <- function(alpha_prior, beta_prior, k_recent, n_recent, k_pr, n_pr) {
  posterior <- posterior_params(alpha_prior, beta_prior, k_recent, n_recent)
  prob <- posterior_predictive(posterior$alpha, posterior$beta, k_pr, n_pr)

  cat("--- Results ---\n")
  cat("Posterior Mean Failure Rate:", posterior$alpha / posterior$beta, "\n")
  cat("Posterior Predictive Probability of", k_pr, "failures in", n_pr, "tests:", prob, "\n\n")
}

#alpha_prior = Historical failures
#beta_prior = Historical total tests
#k_recent = Recent failures
#n_recent = Recent total tests
#k_pr = Pull request failures
#n_pr = Pull request total tests
print("Significant historical, limited mixed results in PR, possible on-going issue")
analyze_scenario(
  alpha_prior = 10,
  beta_prior = 1000,
  k_recent = 7,
  n_recent = 27,
  k_pr = 1,
  n_pr = 2
)

print("Limited historical, limited mixed results in PR")
analyze_scenario(
  alpha_prior = 1,
  beta_prior = 30,
  k_recent = 0,
  n_recent = 5,
  k_pr = 1,
  n_pr = 2
)

print("Limited historical, unlikely regression in PR")
analyze_scenario(
  alpha_prior = 1,
  beta_prior = 30,
  k_recent = 0,
  n_recent = 20,
  k_pr = 1,
  n_pr = 10
)

print("Limited historical, obvious regression in PR")
analyze_scenario(
  alpha_prior = 1,
  beta_prior = 30,
  k_recent = 0,
  n_recent = 20,
  k_pr = 10,
  n_pr = 15
)

print("Strong high pass rate historical data, but this test is failing outside our PR in recent runs")
analyze_scenario(
  alpha_prior = 0,
  beta_prior = 1000,
  k_recent = 20,
  n_recent = 30,
  k_pr = 1,
  n_pr = 3
)

# https://sippy.dptools.openshift.org/sippy-ng/component_readiness/test_details?Aggregation=none&Architecture=amd64&Architecture=amd64&FeatureSet=default&FeatureSet=default&Installer=ipi&Installer=ipi&LayeredProduct=none&Network=ovn&Network=ovn&NetworkAccess=default&Platform=gcp&Platform=gcp&Procedure=none&Scheduler=default&SecurityMode=default&Suite=unknown&Suite=unknown&Topology=ha&Topology=ha&Upgrade=none&Upgrade=none&baseEndTime=2024-12-27%2023%3A59%3A59&baseRelease=4.18&baseStartTime=2025-01-17%2000%3A00%3A00&capability=Other&columnGroupBy=Architecture%2CNetwork%2CPlatform%2CTopology&component=Installer%20%2F%20openshift-installer&confidence=95&dbGroupBy=Platform%2CArchitecture%2CNetwork%2CTopology%2CFeatureSet%2CUpgrade%2CSuite%2CInstaller&environment=amd64%20default%20ipi%20ovn%20gcp%20unknown%20ha%20none&flakeAsFailure=0&ignoreDisruption=1&ignoreMissing=0&includeMultiReleaseAnalysis=1&includeVariant=Architecture%3Aamd64&includeVariant=CGroupMode%3Av2&includeVariant=ContainerRuntime%3Acrun&includeVariant=ContainerRuntime%3Arunc&includeVariant=FeatureSet%3Adefault&includeVariant=Installer%3Aipi&includeVariant=Installer%3Aupi&includeVariant=Network%3Aovn&includeVariant=Owner%3Aeng&includeVariant=Platform%3Aaws&includeVariant=Platform%3Aazure&includeVariant=Platform%3Agcp&includeVariant=Platform%3Ametal&includeVariant=Platform%3Avsphere&includeVariant=Topology%3Aha&includeVariant=Topology%3Amicroshift&minFail=3&passRateAllTests=0&passRateNewTests=95&pity=5&sampleEndTime=2025-01-24%2023%3A59%3A59&samplePRNumber=&samplePROrg=&samplePRRepo=&sampleRelease=4.18&sampleStartTime=2025-01-17%2000%3A00%3A00&testBasisRelease=4.17&testId=cluster%20install%3A3e14279ba2c202608dd9a041e5023c4c&testName=install%20should%20succeed%3A%20infrastructure&view=
print("Slight Regression found from CR recently and then 3 failures out of 10 in a PR.")
analyze_scenario(
  alpha_prior = 0,
  beta_prior = 418,
  k_recent = 8,
  n_recent = 106,
  k_pr = 3,
  n_pr = 10
)

File: pkg/util/bayes/negative_binomial.go
package main

import (
	"fmt"
	"math"
)

// Calculate the posterior parameters
func posteriorParams(alphaPrior, betaPrior, kRecent, nRecent float64) (float64, float64) {
	alphaPosterior := alphaPrior + kRecent
	betaPosterior := betaPrior + nRecent
	return alphaPosterior, betaPosterior
}

// Factorial function
func factorial(n int) float64 {
	if n == 0 {
		return 1
	}
	result := 1.0
	for i := 1; i <= n; i++ {
		result *= float64(i)
	}
	return result
}

// Binomial coefficient function
func binomialCoefficient(n, k int) float64 {
	if k > n {
		return 0
	}
	return factorial(n) / (factorial(k) * factorial(n-k))
}

// Negative Binomial probability calculation
func negativeBinomial(k int, size, prob float64) float64 {
	chooseFactor := binomialCoefficient(k+int(size)-1, k)
	return chooseFactor * math.Pow(prob, size) * math.Pow(1-prob, float64(k))
}

// Calculate the posterior predictive probability
func posteriorPredictive(alphaPosterior, betaPosterior, k, n float64) float64 {
	prob := betaPosterior / (betaPosterior + n)
	return negativeBinomial(int(k), alphaPosterior, prob)
}

// Analyze a scenario
func analyzeScenario(alphaPrior, betaPrior, kRecent, nRecent, kPr, nPr float64) {
	// Step 1: Update posterior parameters
	alphaPosterior, betaPosterior := posteriorParams(alphaPrior, betaPrior, kRecent, nRecent)

	// Step 2: Calculate posterior predictive probability for pull request data
	prob := posteriorPredictive(alphaPosterior, betaPosterior, kPr, nPr)

	// Step 3: Report results
	fmt.Println("--- Results ---")
	fmt.Printf("Posterior Mean Failure Rate: %.4f\n", alphaPosterior/betaPosterior)
	fmt.Printf("Posterior Predictive Probability of %.0f failures in %.0f tests: %.6f\n\n", kPr, nPr, prob)
}

func main() {
	fmt.Println("Significant historical, limited mixed results in PR, possible ongoing issue")
	analyzeScenario(10, 1000, 7, 27, 1, 2)

	fmt.Println("Limited historical, limited mixed results in PR")
	analyzeScenario(1, 30, 0, 5, 1, 2)

	fmt.Println("Limited historical, unlikely regression in PR")
	analyzeScenario(1, 30, 0, 20, 1, 10)

	fmt.Println("Limited historical, obvious regression in PR")
	analyzeScenario(1, 30, 0, 20, 10, 15)

	fmt.Println("Strong high pass rate historical data, but this test is failing outside our PR in recent runs")
	analyzeScenario(0, 1000, 20, 30, 1, 3)

	fmt.Println("Slight Regression found from CR")
	analyzeScenario(0, 418, 8, 106, 3, 10)
}

